### PR TITLE
RDBC-864 Node.js client bulk insert performance

### PR DESCRIPTION
**Pull request by ml054**
https://issues.hibernatingrhinos.com/issue/RDBC-864/Node.js-client-bulk-insert-performance

### CODE REVIEW

File: src/Documents/BulkInsert/BulkInsertWriterBase.ts [hunk 1]
```diff
@@ -51,8 +51,12 @@ export class BulkInsertWriterBase implements IDisposable {
         this.onCurrentWriteStreamSet(this._currentWriter);
     }
 
+    public isFlushNeeded() {
+        return this._currentWriter.length > this._maxSizeInBuffer;
+    }
+
     public async flushIfNeeded(force = false): Promise<void> {
-        if (this._currentWriter.length > this._maxSizeInBuffer || this._asyncWriteDone) {
+        if (this.isFlushNeeded() || this._asyncWriteDone) {
```
Comment (ayende) [inline src/Documents/BulkInsert/BulkInsertWriterBase.ts]: I think that `this._asyncWriteDone` should be in `isFlushNeeded`

File: src/Documents/BulkInsertOperation.ts [hunk 2]
```diff
@@ -553,6 +553,28 @@ export class BulkInsertOperation extends BulkInsertOperationBase<object> {
...
     }
 
+    public tryStoreSync(entity: object, id: string): boolean;
+    public tryStoreSync(entity: object, id: string, metadata: IMetadataDictionary): boolean;
+    public tryStoreSync(
+        entity: object,
+        id: string,
+        metadata?: IMetadataDictionary): boolean {
+
+        if (this.isFlashNeeded() || this._first) {
```
Comment (ayende) [inline src/Documents/BulkInsertOperation.ts]: `isFlushNeeded`, no?

File: src/Documents/BulkInsertOperation.ts [hunk 3]
```diff
@@ -851,6 +878,10 @@ export class BulkInsertOperation extends BulkInsertOperationBase<object> {
         }
     }
 
+    private isFlashNeeded(): boolean {
```
Comment (ayende) [inline src/Documents/BulkInsertOperation.ts]:
```suggestion
    private isFlushNeeded(): boolean {
```

File: src/Documents/BulkInsertOperation.ts (hunk 2 above)
Comment (ml054) [reply src/Documents/BulkInsertOperation.ts]: no, first guarantees that stream was ensured and it was initialized.

File: src/Documents/BulkInsert/BulkInsertWriterBase.ts (hunk 1 above)
Comment (ml054) [reply src/Documents/BulkInsert/BulkInsertWriterBase.ts]: yes. It generates more falses in tryStoreSync, but we wait less overall.

**Review by ayende — APPROVED** [review]

**Comment by ayende** [comment]
Trying to run with this gives me:

```
  [cause]: ReferenceError: fetch is not defined
      at eu.send (C:\Work\tmp\node_modules\ravendb\dist\cjs\ravendb.cjs:5:4667)
      at o._send (C:\Work\tmp\node_modules\ravendb\dist\cjs\ravendb.cjs:5:44540)
      at async o._sendRequestToServer (C:\Work\tmp\node_modules\ravendb\dist\cjs\ravendb.cjs:5:43256)
      at async o._executeOnSpecificNode (C:\Work\tmp\node_modules\ravendb\dist\cjs\ravendb.cjs:5:41974)
      at async o._handleServerDown (C:\Work\tmp\node_modules\ravendb\dist\cjs\ravendb.cjs:5:51434)
      at async o._sendRequestToServer (C:\Work\tmp\node_modules\ravendb\dist\cjs\ravendb.cjs:5:44316)
      at async o._executeOnSpecificNode (C:\Work\tmp\node_modules\ravendb\dist\cjs\ravendb.cjs:5:41974)
      at async bs._getNextRange (C:\Work\tmp\node_modules\ravendb\dist\cjs\ravendb.cjs:10:28284)
      at async bs.getNextId (C:\Work\tmp\node_modules\ravendb\dist\cjs\ravendb.cjs:10:28051)
      at async bs.generateDocumentId (C:\Work\tmp\node_modules\ravendb\dist\cjs\ravendb.cjs:10:27699)
```

* checked out v6.0
* pulled your changes
* npm run build, npm pack
* npm i /path/to/ravendb-6.0.0.tgz

```
 const store = new DocumentStore('http://localhost:8080', 'bulk');
    store.initialize();
    const bulk = store.bulkInsert();
    for (let i = 0; i < 100_000_000; i++) {
        const user = new User('user' + i);
        const id = "users/" + i;
        if (bulk.tryStoreSync(user, id) == false) {
            await bulk.store(user, id);
        }
    }
    console.log('done storing');
    await bulk.finish();
```

Tried with and without `tryStoreAsync`, same error.

**Comment by ayende** [comment]
When running in debug, it complains about missing `ReadableStream` and then about `fetch`.
```
$ node --version
v16.14.0
```
Upgraded to node v20 and it appears to be working

**Comment by ml054** [comment]
in 6.0 we require node.js 18+ 
https://github.com/ravendb/ravendb-nodejs-client/blob/v6.0/package.json#L40

however node 16 doesn't seem to respect this setting. :(

**Pull request merged**